doc/UsersGuide/source/ContributingDevelopment.rst

```diff
@@ -13,4 +13,284 @@ Contributing Development
+
+In addition to these libraries, other software is required such as a compiler (Intel or GNU), MPI library (impi), HDF5, cmake, Python, Perl and Ruby to compile the model and run regression tests.
+
+--------------------------------------------
+Making code changes using a forking workflow
+--------------------------------------------
+
+If developers would like to make code changes, they need to make a personal fork, set up upstream remote (for merging with the original ufs-weather-model), and create a branch for ufs-weather-model and each of the subcomponent repositories they want to change. They can then make code changes, perform testing and commit the changes to the branch in their personal fork. It is suggested that they merge their branch with the develop branch of the original repositories periodically to get the latest updates and bug fixes.
+
+If developers would like to get their code committed back to the original repository, it is suggested to follow the steps below:
+
+      # Create an issue in the authoritative repository. For example to commit code changes to fv3atm, please go to https://github.com/NOAA-EMC/fv3atm, under NOAA-EMC/fv3atm and find the “Issues” tab next to the “Code” tab. Click on “Issues” and a new page will appear. On the right side of the page, there is a green “New issue” button. Clicking on that will lead to a new issue page. Fill out the title, comments to describe the code changes, and also please provide personal fork and branch information. Lastly, click on the “Submit new issue” button, so that the new issue is created.
+
+      # When the development is mature, tests have been conducted, and the developer is satisfied with the results, create a pull request to commit the code changes.
+
+      * Merge developer’s branch to the latest ufs-weather-model develop branch in authoritative repository. If changes are made in model sub-components, developers need to merge their branches to branches with the corresponding authoritative repository (or original repository for some components). For this, code management practices of the subcomponents need to be followed.
+
+      * Regression tests need to pass on at least one supported platform.
+
+      * For each component branch where developers make changes, developers need to go to their personal fork on GitHub and click on the “New pull request” button. When a new page “Compare changes” appears, developers will choose the branch in their fork with code changes to commit and the branch in upstream repository that the changes will be committed to. The code differences between the two branches will be displayed. Developers can review the differences and click on “submit pull request” to make the pull request. After code changes are committed to the component repository, developers will make pull requests to ufs-weather-model repository.
```

Comment:
Please consider including info from Dom and other code managers on procedure for nested PRs in the UFS. Rewrite as needed for clarity.

(1) When PRs are created, the creator must temporarily modify .gitmodules to point to his/her fork and branch if updates are required for submodules.

(2) Merging code from PRs with submodules requires coordination with the person making the PRs. From the "innermost" nested PR up to the top-level PR, the PRs need to be merged as-is. After each merge, the person creating the PRs has to update his/her local code to check out the merged version, revert the change to .gitmodules, and push this to GitHub to update the PR. And so on and so forth.

(3) Checking out the code ufs_release_1.0 should always be as follows:
git clone https://github.com/NOAA-EMC/NCEPlibs
cd NCEPlibs
git checkout ufs_release_1.00
git submodule update --init --recursive

(4) Checking out a PR with id ID for testing it should always be as follows:
git clone https://github.com/NOAA-EMC/NCEPlibs
cd NCEPlibs
git fetch origin pull/ID/head:BRANCHNAME
git checkout BRANCHNAME
git submodule update --init --recursive

For my current NCEPLIBS PRs, this is simple: first, accept and merge the cmake PR - which you did. Then let me update cmake to point back to NOAA-EMC in all repositories that use the cmake modules repository as submodule - which I did. Then merge the NCEPLIBS-xyz PRs (can be done in parallel) - hasn't been done yet. Then let me update my NCEPLIBS umbrella PR to point the submodules back to NOAA-EMC. Then merge this one. Done.

Reply:
Done